[draft] AIP-X Disable safe burn of user owned objects (#492)
aptos-foundation/AIPs: aips/aip-x-disable-object-burn.md

---
aip: (this is determined by the AIP Manager, leave it empty when drafting)
title: Disable Object Burn
author: gregnazario (greg@aptoslabs.com)
discussions-to (*optional): <a url pointing to the official discussion thread>
Status: Draft
last-call-end-date (*optional): <mm/dd/yyyy the last date to leave feedbacks and reviews>
type: Standard (Framework)
created: 8/28/2024
updated (*optional):
requires (*optional):
---

# AIP-X - Disable Safe Burn of User-Owned Objects

## Summary

Safe burn was originally described in [AIP-45](./aip-45.md), which allowed users to tombstone their objects and remove
direct linkage of an undeletable, also known as *soulbound*, object to an account. This AIP seeks to disable safe
object burn, as it caused extra complexity, and sometimes unexpected consequences. Owners of burned objects will still

> be able to unburn and recover their tombstoned objects.

Reviewer: especially with the desire to create untransferable objects and the notion of ownership which becomes muddy when the burn address manages the object but the object is still owned by another address.

### Goals

> As a result of this AIP, users will still be able to unburn their burnt objects, but will not be able to burn any new
> undeletable objects. This will reduce complexity and some interesting edge cases associated.

Reviewer: ```suggestion
As a result of this AIP, users will still be able to unburn their burnt objects, but will not be able to burn any new objects. This will reduce complexity and some interesting edge cases associated.
```

> If we delay enacting this, there may be more burnt objects, which could cause more unexpected behaviors around moving

Reviewer: ```suggestion
If we delay enacting this, there may be more burnt objects, which increase risk to new protocols due to the nuanced behaviors around burnt objects.
```

> burnt objects to the burn address.

Reviewer: ```suggestion
```

### Out of Scope

This removes the previous mitigation, since users cannot opt-out of receiving unsolicited content. Wallets or other
platforms will need to come up with a standardized way of hiding unwanted items.

Additionally, we will not be preventing unburning existing burnt objects at this time.

> ## Motivation

Reviewer 2: The motivation here is not so clear to me. It mostly mentioned the complexity introduced, but is there a specific example to help understand this complexity better?

Author: This leaves room for multiple vulnerabilities, that I did not want to mention.

The purpose here is to remove burn to prevent confusion, and complexity around *soulbound* objects changing owners.
Originally, this was meant mostly for being able to hide unwanted *soulbound* NFTs. But, there are more use cases for
*soulbound* objects (e.g. primary fungible store), which can complicate functionality around them. Removing this
functionality simplifies cases for builders, and can make it easier to accomplish things like allowlists on fungible
assets.

If we do not accept this proposal, some functionality around *soulbound* objects will require extra complexity or even
may not be able to work around burns.

## Impact

This impacts the following parties:

> * Users who have already burnt, or plan to burn owned *soulbound* objects

Reviewer 2: can we show some data on how many burns happened in the past to understand the impact better?

Author: I do not have this data as it is not indexed.

The complexity of handling indexing for something that was purely driven because a couple engineers didn't like having soulbound NFTs in their accounts is low.

It's not handled by any exchange, or by any NFT tooling, and is barely documented.

Reviewer: how does it impact those that have already burnt?

Author: It does not impact already burnt objects.  However, objects that are burnt can still be unburnt.

Let me add that.

NFTs. Additionally, if any smart contracts built around the `burn` function, it is possible they would stop working.

However, given that `burn` is just essentially moving an object to a different address with a Tombstone, there should be
nearly no risk associated with removing this functionality.

This still allows users to `unburn` already burnt objects, meaning that users will still have access to items they had
burnt previously.

## Security Considerations

The result of this AIP should strictly be more secure than allowing `burn`s. Burns move objects to be owned by a burn
address, which can cause some unwanted side effects around allowlisting by owner. By ensuring that the object is always
owned by the original owner, it should mitigate any issues there.

## Future Potential

The future is that wallets should provide their own allowlisting of which assets to show, or provide a decentralized
allowlist of which assets to show in a wallet.

## Timeline

### Suggested implementation timeline

The expectation is that this should ship as part of the next release after being approved.

### Suggested developer platform support timeline

> One day of work, to remove `burn` from any documentation on the [aptos.dev](https://aptos.dev) website.

Reviewer: I think it should just be explicit callout that burn needs to be stricken from aptos.dev not an estimate on the amount of work
